### PR TITLE
Integrate identity keys with libp2p-noise for authentication.

### PR DESCRIPTION
**Pull request by romanb**
## TL;DR
Make `libp2p-noise` usable with a `Swarm` by bridging the gap between static DH keys and identity keys, thereby enabling the use of `libp2p-noise` with any type of identity keypair through the use of signatures when necessary.

## Motivation
In order to make `libp2p-noise` usable with a `Swarm`, which requires a `Transport::Output` that is a pair whose first component is a peer ID identifying the remote, it is necessary to bridge the gap between static DH public keys and public identity keys from which peer IDs are derived. In particular, before being reported to the Swarm as the remote's identity in form of a peer ID, the public identity key of the remote needs to be verified in the sense that it must be established during the handshake that the remote is either in possession of a corresponding secret identity key or that the static DH public key is authentic w.r.t. the public identity key.

## Implementation
Because the DH static keys and the identity keys need not be related, it is generally not possible to derive the public identity key from the DH static public key. It is thus necessary that the public identity keys are exchanged as part of the Noise handshake (unless known a priori), which the Noise protocol accommodates for through the use of handshake message payloads.

The implementation of the existing (IK, IX, XX) handshake patterns is thus changed to send the public identity keys in the handshake payloads of those handshake messages which also transmit the static DH public keys. Additionally, to facilitate the use of any identity keypair with `libp2p-noise`, before being sent, the local static DH public key is signed using the identity keypair and the signature attached, _unless_ the static DH public key is "linked" to the public identity key by other means, e.g. when an Ed25519 identity keypair is (re)used as an X25519 keypair.

In summary, the following use-cases are supported:

  1. Reusing an ed25519 identity keypair as an x25519 keypair. The public identity key is sent to the remote unless already known and no signature is sent. The public identity key still needs to be sent unless known by the remote, because ed25519 to x25519 conversion is not reversible. The remote identifies the local node by comparing the x25519 static public key to the (converted) ed25519 public key.

  2. Using an ed25519, secp256k1 or RSA identity keypair with a distinct static x25519 keypair. The public identity key is sent to the remote unless already known, the static DH public key is signed using the identity keypair and the signature added to the message payload. The remote identifies the local node by verifying the signature.

Importantly, these cases can be combined in the same handshake, i.e. one party may authenticate without a signature and the other with a signature, depending on how either party creates their x25519 static keys.

### Implementation Notes

  * For ease of implementation and because it allows compact expression of the current and any future handshake roundtrip patterns, a `Handshake` future returned from an inbound or outbound upgrade is currently boxed. If that is deemed unacceptable, the previous `rt1` and `rt15` modules could be reintroduced in a slightly simplified form.  See `protocols/noise/src/io/handshake.rs` for details.

  * To preserve the [identity hiding](http://www.noiseprotocol.org/noise.html#identity-hiding) properties of all handshake patterns, use of the `IK` pattern now expects the initiator to know both, the static DH public key and the public identity key of the remote. These public keys may be "the same" if the remote's identity is an ed25519 public key and the remote is known to (re)use the ed25519 keypair for x25519, since the static x25519 public key can then be derived from the known ed25519 public key.

  * `protobuf` was chosen for the encoding of the handshake message payloads, because it is already used in other parts of libp2p for similar purposes and because the embedded encoding of the public identity keys is also protobuf as defined by the libp2p specs.

  * `build_development_transport` was changed to use noise instead of secio unless building for wasm. This implies that the examples now also use noise.

  * The implementation should not preclude future additions of any handshake patterns, i.e. I've tried to keep the updated implementation based on individually reusable components such that is is easy to add new handshakes.

## Future Work

As was previously [pointed out](https://github.com/libp2p/rust-libp2p/pull/972#issuecomment-467224357) by @burdges, there are newer Noise handshake pattern variants that [directly incorporate signatures](https://github.com/noiseprotocol/noise_sig_spec/blob/master/output/noise_sig.pdf) in place of DH with static keys for authentication. These could potentially be an alternative to simplify use-case `2.` from above, which would then not require both static DH keys and signatures, though it would probably preclude mixing use-case `1.` and `2.` in the same handshake (i.e. both parties then either use signatures or static DH keys, depending on the agreed handshake pattern).

Lastly, I haven't investigated the option of using signatures over the handshake hash in payloads on top of unauthenticated Noise handshake patterns and whether such a construction would be equivalent to Noise handshakes with built-in signatures (also over the handshake hash), as I only wanted to add identity key authentication on top of the already implemented Noise handshakes `IK`, `IX` and `XX`, which are based on static DH keys, to make these actually usable with libp2p's `Swarm`. Thus, this is also a direction that could be investigated further.

### CODE REVIEW

File: protocols/noise/src/protocol/x25519.rs [hunk 1]
```diff
@@ -143,6 +151,28 @@ impl Keypair<X25519> {
...
+    ///
+    /// > **Note**: If the identity keypair is already used in the context
+    /// > of other cryptographic protocols outside of Noise, e.g. for
+    /// > signing in the `secio` protocol, it should be preferred to
+    /// > create a new static X25519 keypair for use in the Noise protocol.
+    /// >
+    /// > See also:
+    /// >
+    /// >  * [Noise: Static Key Reuse](http://www.noiseprotocol.org/noise.html#security-considerations)
+    pub fn from_identity(id_keys: &identity::Keypair) -> Keypair<X25519> {
```
[the remark this answers is not in the document]
Comment (romanb) [reply protocols/noise/src/protocol/x25519.rs]: Yes, that is probably better and has less potential for confusion.

File: protocols/noise/src/io/handshake.rs [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+        loop {
+            match mem::replace(&mut self.state, RecvIdentityState::Done) {
+                RecvIdentityState::Init(st) => {
+                    self.state = RecvIdentityState::ReadPayloadLen(nio::read_exact(st, [0u8; 2]));
+                },
+                RecvIdentityState::ReadPayloadLen(mut read_len) => {
+                    if let Async::Ready((st, bytes)) = read_len.poll()? {
+                        let mut len_be = [0u8; 2];
+                        len_be.copy_from_slice(&bytes);
+                        let len = u16::from_be_bytes(len_be) as usize;
```
[the remark this answers is not in the document]
Comment (romanb) [reply protocols/noise/src/io/handshake.rs]: Indeed!

**Comment by tomaka** [comment]
Correct me if I'm wrong, but isn't the point of Noise that we shouldn't change the handshake patterns because they are proven?

**Comment by romanb** [comment]
> Correct me if I'm wrong, but isn't the point of Noise that we shouldn't change the handshake patterns because they are proven?

@tomaka Why do you think this changes the handshake patterns? The handshake patterns are the same, just passing public identity keys and (if necessary) a signature of the static DH public key along in the handshake payloads, in order to link the authenticity of the static DH keys to the libp2p identity keys and thus actually make these connection upgrades usable with a Swarm by being able to produce a `PeerId`. This is pretty much an intended use-case for handshake payloads, see e.g. [Section 14 - Security Considerations - Authentication](http://www.noiseprotocol.org/noise.html#security-considerations).

**Comment by tomaka** [comment]
Isn't the point of the IK handshake, for example, that we know the public key in advance and don't need to share it? And that the point of the XX pattern is that we don't know it in advance?

**Comment by romanb** [comment]
> Isn't the point of the IK handshake, for example, that we know the public key in advance and don't need to share it? And that the point of the XX pattern is that we don't know it in advance?

Yes (to be precise, with `IK` the initiator knows the public key of the responder, not vice versa). Why do you think that changed here?

**Comment by twittner** [comment]
Can we move this forward or are there any more objections? @tomaka? @romanb? @burdges?

**Comment by tomaka** [comment]
My only few concerns are:

- Isn't it better to *somehow* adjust the `Swarm` to always manipulate ED25519 public keys instead of generic `PublicKey`s or `PeerId`s?
- What something else than protobuf, instead of CBOR? It would also make the Rust code much more straight-forward.

However I'd fine to merge this in order to break the status quo of lib2p-noise being unusable.

**Comment by romanb** [comment]
> My only few concerns are:
> 
>     * Isn't it better to _somehow_ adjust the `Swarm` to always manipulate ED25519 public keys instead of generic `PublicKey`s or `PeerId`s?

Do you mean actually restricting the `Swarm` (and libp2p_core in general) to only work with ed25519 keys? I think that sounds very inflexible and would make a graceful transition to new key types impossible. Surely there always needs to be at least the option to temporarily support multiple types of keys (this is similar to what was discussed in https://github.com/paritytech/substrate/pull/1971#issuecomment-472792137. Of course `libp2p_noise` itself could be restricted to be usable only with a subset of the key types of `libp2p_core`, like only ed25519 but I actually thought it to be a good idea to be able to use any identity keypair from `libp2p_core` with `libp2p_noise`, merely by using the signing capabilities that identity keypairs have. Also, the restriction to ed25519 would not alleviate the need to transmit the public identity keys in the handshake payloads alongside the static x25519 public keys, due to losing the sign bit in the ed25519 -> x25519 conversion, so it would not be that much of a simplification in this implementation.

>     * What something else than protobuf, instead of CBOR? It would also make the Rust code much more straight-forward.

As I mentioned in the PR description, I merely chose protobuf because I reused the existing (and libp2p-standardised) protobuf encoding of `PublicKey`s provided by `libp2p_core`, so that the payloads are protobuf-in-protobuf. Using something else, like CBOR, would only make sense to me if we first define a CBOR encoding for `PublicKey`s, which would further expand the scope of this PR.

My suggestion is as follows:

  * I rebase this PR on master to make it mergable and document the current state of `libp2p-noise` as experimental (i.e. it works and can be used with `libp2p-core` as an alternative to `libp2p-secio` but is not recommended for production use, due to potential breaking changes to both API and the wire protocol).

  * If this implementation should survive longer than anticipated and it is actually supposed to be released as a usable, stable first version before it gets replaced by a "v2" based on noise-explorer-generated code, we can reconsider the choice of encoding by e.g. by moving to CBOR.

How does that sound?

**Comment by tomaka** [comment]
> How does that sound?

Sounds good to me!

> Using something else, like CBOR, would only make sense to me if we first define a CBOR encoding for PublicKeys, which would further expand the scope of this PR.

What I was thinking is that the public key would be transmitted as a protobuf binary blob, but the "identity" message would be some CBOR struct.

There are two reasons for that:

- It would make the Rust code much cleaner. Using CBOR can be done with a serde derive attribute on a struct, as opposed to a `.proto` file that is turned into Rust through a script.

- Protocol Labs has been switching to CBOR in Filecoin and in general orient themselves towards CBOR for newer protocols.

**Comment by burdges** [comment]
> Do you mean actually restricting the Swarm (and libp2p_core in general) to only work with ed25519 keys? I think that sounds very inflexible and would make a graceful transition to new key types impossible.

A priori, I'd think Ed25519 sounds sufficient until we want post-quantum authentication really, but yeah maybe an etehreum bridge parachain might need secp256k1 or something.

**Comment by romanb** [comment]
> > Using something else, like CBOR, would only make sense to me if we first define a CBOR encoding for PublicKeys, which would further expand the scope of this PR.
> 
> What I was thinking is that the public key would be transmitted as a protobuf binary blob, but the "identity" message would be some CBOR struct.
> 
> There are two reasons for that:
> 
>     * It would make the Rust code much cleaner. Using CBOR can be done with a serde derive attribute on a struct, as opposed to a `.proto` file that is turned into Rust through a script.
> 
>     * Protocol Labs has been switching to CBOR in Filecoin and in general orient themselves towards CBOR for newer protocols.

I see. I'm generally in favor of CBOR as well, I just thought protobuf-in-cbor is unnecessarily cumbersome, since you now have to parse two different encodings with two different libraries, but its certainly an option. I'm glad to hear that Protocol Labs is generally favoring CBOR over protobuf now.

I updated the branch as discussed and will merge it now.
